### ISSUE TITLE
Citing Experimental RFCs as if being a part of CUBIC

### ISSUE THREAD

**Issue by goelvidhi** [NTAP/rfc8312bis#90]
Markku Kojo said,
> The draft says that CUBIC MAY implement DSACK [RFC3708], limited slow
  start [RFC 3742], [RFC7661] and hybrid slow start [cites a paper].
  Aren't the first three down references? Not sure if it is appropriate
  for a Stds Track document to cite experimental work or a paper like
  this even though it's a MAY.

**Comment by larseggert** [comment]
These references are all **informational**, which means they are not DOWNREFs by definition (see [RFC3967)](https://datatracker.ietf.org/doc/html/rfc3967).

Propose to close with no action. Markku?

**Comment by larseggert** [comment]
No word from @markkukojo, closing this. @markkukojo, please reopen if you disagree.

**Comment by markkukojo** [comment]
Right, the references are currently informal but I think the issue is more subtle and how some of the experimental RFCs are now cited puts them into a "grey area" .

Sec 4.10:
In sec 4.10 the draft says "CUBIC MUST employ slow-start" and then continues by allowing  CUBIC to use RFC 3742 or [HR08] as the slow start algorithm for fast and long-distance networks by`using the keyword MAY. So, an implementer MUST implement slow- start and if he/she selects to implement RFC 3742 he/she must read and understand RFC 3742 in order to implement it. In my view it would require RFC 3742 to become a normative reference and thereby a down ref, or at least puts it in a grey area.

A solution to this could be simply to say:    "CUBIC MUST employ a slow-start algorithm [RFC 5681]."
And it does not need to say anything like "MAY use x" because RFC 3742 (and [HR08]) is less aggressive than regular slow start so use of it is allowed as any other experimental RFC (for experimental purposes). 

I understand that the problem here is that if one decides to use [HR08], one needs to take that into account when setting _W_max_ , etc. as described in the draft. However, I'm not sure whether it's reasonable to allow (by using MAY) employing an algorithm in a research paper (or in an Exp RFC) as part of a Stds Track RFC? My interpretation of using MAY in this way is that it sends a message that "while the cited algorithm is experimental (or research) only and should not be used in the public Internet other than for experimental purposes, it is, however,  ALL OK to deploy it when used with CUBIC".
 
So, instead of using MAY, the draft could simply say that "if CUBIC uses [HR08], it may exit ..."
This suggests dropping the second sentence (starting: "Among ...") of the first para as unnecessary and potentially confusing. 

Sec 4.9:

Here also the use of MAY is quite unnecessary and potentially results in similar confusion about the status of RFC 3708.
There is no reason to separately allow (by using MAY) CUBIC to employ RFC 5682 or RFC 3522 as they are Stds Track, nor to separately allow employing RFC 3708 as it can be employed with same provisions as any Experimental algorithm.  

Again, it would be just enough to note that if any of these are used, the mentioned CUBIC variables need to be restored as described. However, the authors should carefully consider what RFC 4015 says and what limitations there are for a congestion control response when a spurious RTO is detected. RFC 4015 is currently the only Stds Track CC response algorithm for spurious RTOs. There is no Stds Track CC response algorithm for false Fast Retransmits; this topic was studied around two decades ago and there was no consensus  on whether restoring CC state is a good idea in all scenarios. E.g., in mobile networks, a make-before-break handoff from a high RTT path to a low RTT path may result in packet reordering when the available network capacity actually decreases and restoring the old state in such a case is not a sane action.  Therefore, I think that this draft is not in position to provide a reasonable suggestion to unconditionally restore CC state after spurious congestion events nor to provide evidence that it is always the correct action. Moreover, handling spurious retransmissions is not a key part of CUBIC but it is a more general issue where other CCs have the same challenges as CUBIC and therefore I think the CC response for spurious rexmits should be handled separately and should follow common principles for all CCs. 

Sec 4.6:

The draft allows using RFC 7661 as described in the RFC. However, if I recall it correctly, the equations in RFC 7661 assume that cwnd increase and decrease uses standard methods i.e., alfa=1 and beta=0.5. If the equations in RFC  7661 are directly applied with CUBIC, the outcome might not be expected nor correct. I didn't have time to check this, so I might be wrong, but this is good to double-check with Gorry and/or other authors of RFC 7661 who have better insight to the details of RFC 7661.

**Comment by larseggert** [comment]
Please check #90, which intends to address your first two comments.

About RFC7661, would @gorryfair please chime in?

**Comment by goelvidhi** [comment]
@markkukojo 
> Therefore, I think that this draft is not in position to provide a reasonable suggestion to unconditionally restore CC state after spurious congestion events nor to provide evidence that it is always the correct action. Moreover, handling spurious retransmissions is not a key part of CUBIC but it is a more general issue where other CCs have the same challenges as CUBIC and therefore I think the CC response for spurious rexmits should be handled separately and should follow common principles for all CCs.

Response to spurious events due to Fast retransmits was added newly to 8312-bis draft as it made sense to cover this scenario which is likely more common than spurious RTOs and folks who implemented QUIC noticed a great throughput improvement when implementing this. I agree that this is common to all CCs but at as of today, I don't think any RFC covers it. For completeness, it made sense to add it to CUBIC.

> There is no Stds Track CC response algorithm for false Fast Retransmits; this topic was studied around two decades ago and there was no consensus on whether restoring CC state is a good idea in all scenarios. E.g., in mobile networks, a make-before-break handoff from a high RTT path to a low RTT path may result in packet reordering when the available network capacity actually decreases and restoring the old state in such a case is not a sane action.

How common is the scenario you are describing above? And if we are switching to low RTT, then the sender will soon-after reduce the congestion window if restoring to old state was a mistake. (i.e. if the old state cwnd > capacity for low RTT path)

**Comment by goelvidhi** [comment]
>The draft allows using RFC 7661 as described in the RFC. However, if I recall it correctly, the equations in RFC 7661 assume that cwnd increase and decrease uses standard methods i.e., alfa=1 and beta=0.5. If the equations in RFC 7661 are directly applied with CUBIC, the outcome might not be expected nor correct. I didn't have time to check this, so I might be wrong, but this is good to double-check with Gorry and/or other authors of RFC 7661 who have better insight to the details of RFC 7661.

Regarding RFC 7661, there are two things to consider,
1. Threshold for validated vs non-validated phase - this threshold has been set by 7661 based on slow-start increase and is valid for CUBIC too.
> A standard TCP sender in slow-
   start is permitted to double its FlightSize from one RTT to the next.
   This motivated the choice of a threshold value of 1/2. This threshold ensures a
    sender does not further increase the cwnd as long
   as the FlightSize is less than (1/2*cwnd)

2. Congestion response - RFC 7661 uses 1/2 as the reduction factor which is probably taken from New Reno. I think this is what we need to update in CUBIC. Something like this,

```
CUBIC updates the response to congestion when RFC 7661 is used to validate the congestion window. It updates the congestion window for loss recovery as below,

cwnd = (Max(pipeACK,LossFlightSize)) * Beta_cubic

It also updates the congestion window at the end of the recovery phase to,
cwnd = (Max(pipeACK,LossFlightSize) - R) * Beta_cubic
```

I don't know how to best document this, should we re-write these equations and explain all the variables or is there a way to reference RFC 7661 and just say, If you use RFC 7661, CUBIC updates Section 4.4.1 of RFC 7661 to replace the 0.5 reduction rate by Beta_cubic?

**Comment by larseggert** [comment]
The latter seems easier to me, but I don't have a strong preference. Waiting to hear from @markkukojo and/or @gorryfair.

**Comment by larseggert** [comment]
I'd like to merge #111 soon. Is there anything that needs to be changes in it to address this issue?

**Comment by markkukojo** [comment]
I'll take a look at it later today.

**Comment by markkukojo** [comment]
> @markkukojo
> 
> > Therefore, I think that this draft is not in position to provide a reasonable suggestion to unconditionally restore CC state after spurious congestion events nor to provide evidence that it is always the correct action. Moreover, handling spurious retransmissions is not a key part of CUBIC but it is a more general issue where other CCs have the same challenges as CUBIC and therefore I think the CC response for spurious rexmits should be handled separately and should follow common principles for all CCs.
> 
> Response to spurious events due to Fast retransmits was added newly to 8312-bis draft as it made sense to cover this scenario which is likely more common than spurious RTOs and folks who implemented QUIC noticed a great throughput improvement when implementing this. I agree that this is common to all CCs but at as of today, I don't think any RFC covers it. For completeness, it made sense to add it to CUBIC.
> 

Let me try again. There are several reasons why the existence of the current text is problematic:
1) Spurious rexmits are not part of CUBIC nor specific to CUBIC. I think this draft should focus on CUBIC and not include various features present in Linux or any other stack, no matter how useful they are particularly when the problem space is orthogonal to CUBIC.  UNDO of congestion control state was there in Linux well before CUBIC was introduced and before the cited RFCs were published.
2) As I said, such an UNDO comes with caveats. Certainly it provides benefits in certain scenarios, but may also be harmful in some other scenarios, particularly if various safeguards are not implemented. 3) Why there is no RFC to cover response for false fast retransmits? Because there are cases when one should not revert the CC state as it would result in undesired behaviour, such as incorrectly bursting uncontrolled data in case one should actually reduce cwnd. This is also the reason why all these detection algorithms (RFC 5682, RFC 3522, and RFC 3708) were separated from the response. 
4) The way that current text proposes for reverting the congestion control state in case of a spurious RTO is in conflict with RFC 4015 which is Stds Track. What's wrong in the steps specified in RFC 4015 such that one should not follow it but instead follow this draft which in turn would require an update to  RFC 4015? 
If this draft is in need to say anything, it should cite RFC 4015 for response to detected spurious RTOs and indicate the additional state information to store and restore in case of CUBIC.

> > There is no Stds Track CC response algorithm for false Fast Retransmits; this topic was studied around two decades ago and there was no consensus on whether restoring CC state is a good idea in all scenarios. E.g., in mobile networks, a make-before-break handoff from a high RTT path to a low RTT path may result in packet reordering when the available network capacity actually decreases and restoring the old state in such a case is not a sane action.
> 
> How common is the scenario you are describing above? And if we are switching to low RTT, then the sender will soon-after reduce the congestion window if restoring to old state was a mistake. (i.e. if the old state cwnd > capacity for low RTT path)

The scenario above is just one example. I don't have an answer how common it might be today. AFAIK at least cellular networks used to attempt delivering data in order up until 3G. Don't know that well what is the exact situation with 4G and 5G (or 6G).  But if one would like to propose reverting CC state in case of false fast retransmits one should provide evidence that such scenarios are non-existing today and unlikely to (re)appear in the near future. E.g., is there evidence that route fluttering does not exist anymore? Moreover, there have recently been proposals which suggest that link layers should not anymore deliver data in order. E.g. Bob has been advocating it. How about if it gets realised anytime soon with some technologies?

Sure the sender will reduce the congestion window if restoring to old state was a mistake, but the new available capacity might be only a small fraction of the old and it would possibly require reducing cwnd a notable number of times before cwnd decreases enough. And such a sender behaviour would be detrimental to other flows from which it would steal the bandwidth unfairly.

**Comment by markkukojo** [comment]
> 1. Congestion response - RFC 7661 uses 1/2 as the reduction factor which is probably taken from New Reno. I think this is what we need to update in CUBIC. Something like this,
> 
> ```
> CUBIC updates the response to congestion when RFC 7661 is used to validate the congestion window.
> It updates the congestion window for loss recovery as below,
> 
> cwnd = (Max(pipeACK,LossFlightSize)) * Beta_cubic
> 
> It also updates the congestion window at the end of the recovery phase to,
> cwnd = (Max(pipeACK,LossFlightSize) - R) * Beta_cubic
> ```

Has this been implemented for CUBIC and experimented with so we know it works properly? And do we have data to show this is the case?

In other words, would this be mature enough to become Stds Track and just for CUBIC, while RFC 7661 is experimental and the problem space once again is not specific to CUBIC but common for all CCs?

Anyway, I'd like to hear from those who have worked with RFC 7661 and do have a lot of experience on it. And would definitely have much better insight in any potential caveats that might be buried in there.

> I don't know how to best document this, should we re-write these equations and explain all the variables or is there a way to reference RFC 7661 and just say, If you use RFC 7661, CUBIC updates Section 4.4.1 of RFC 7661 to replace the 0.5 reduction factor by Beta_cubic?

IMO this problem space is common to all CCs and still experimental, so the proposal is not mature enough and best handled in a separate document, either RFC7661bis or a draft of its own that may potentially be incorporated in RFC7661bis later.

**Comment by goelvidhi** [comment]
@huitema @nealcardwell In this issue #23, we discussed and added a new section for Spurious detection and response for CUBIC but @markkukojo doesn't think we should do that due to any adverse effects of reverting to old_cwnd in case there is route change that has much lower capacity than old_cwnd state.

Do you think citing RFC 4015 for spurious RTOs is enough? This would mean we wouldn't provide any guidance for spurious detection of false fast retransmits?

**Comment by huitema** [comment]
It has been experimented -- was the default option for using cubic in picoquic, before switching the default to BBR. No negative feedback. There were also extensive simulations done by the MS Quic team, as part of their systematic unit tests. We should ask them, but I believe these simulations are based on scenarios they have encountered in real deployments. Implementing spurious loss recovery in QUIC-Cubic was key to achieving parity with TCP-Cubic in these scenarios.

Timers are not perfect, spurious timers do happen -- that's quite visible on Wi-Fi and LTE links. Cubic is not performing well in practice without this spurious loss recovery. If this is not specified, implementations will probably just move to use BBR instead of Cubic. There is a perverse bad angel wispering to me that this would be a good thing, since BBR does result in smaller queues all over the Internet. Which means I a bit conflicted...

If you are concerned about standardization status, one possibility would be to move that text to a non normative appendix.

**Comment by nealcardwell** [comment]
> @huitema @nealcardwell In this issue #23, we discussed and added a new section for Spurious detection and response for CUBIC but @markkukojo doesn't think we should do that due to any adverse effects of reverting to old_cwnd in case there is route change that has much lower capacity than old_cwnd state.
>

I would strongly urge the draft to keep the "Spurious Congestion Events" section, since reordering and spurious retransmits are common and getting more common, so undo of CC reductions is critical and becoming more so. 

In the scenario mentioned that has a route change that causes a reordering event and reduction in available bandwidth, it's still fine to undo based on detecting the spurious retransmits. If there are real losses at any point, those will be detected and the CC will react. The likely order in this scenario would be:

- route change to path with lower RTT and bandwidth
- RTT decreases, sender receives SACKs for reordered packets
- enter spurious fast recovery -> spurious retransmit and CUBIC reduces ssthresh and cwnd
- spurious retransmit detected -> CUBIC undo, restoring parameters from before spurious recovery
- real packet losses happen due to lower bandwidth   -> enter well-founded fast recovery -> CUBIC reduces ssthresh and cwnd

> Do you think citing RFC 4015 for spurious RTOs is enough? This would mean we wouldn't provide any guidance for spurious detection of false fast retransmits.

I would suggest citing all the major mechanisms that major TCP stacks use in the real world to detect spurious retransmits, namely something like: "CUBIC implementations MAY use any of the following documented mechanisms for detecting spurious congestion events:  Eifel [rfc3522] [rfc4015], F-RTO [rfc5682], DSACK [rfc3708]."
 
I believe their statuses are:
- Eifel rfc3522 (exp), rfc4015 (std)
- F-RTO: rfc5682 (std)
- DSACK: rfc3708 (exp)
 
Really TCP stacks SHOULD use these, but if there's a problem with a stds track doc saying a stack SHOULD use an experimental RFC then perhaps MAY is OK?

**Comment by larseggert** [comment]
If we want to say "MUST implement" for something that is not on the standards track, maybe that's a reason to republish that document as a PS. Which documents specifically would this be the case for?

**Comment by goelvidhi** [comment]
I was wondering why is Eifel detection (RFC 3522) experimental while the response on detection (RFC 4015) a stds track. 
F-RTO is stds track detection of spurious RTO which does not require timestamps to be enabled (unlike RFC 3522)

Should we say,
`MUST implement F-RTO detection and RFC 4015 response algorithm? MAY implement other algorithms for detection, i.e. RFC 3522 and RFC 3708`

**Comment by larseggert** [comment]
IIRC, RFC3522 has an IPR disclosure that makes it difficult to implement in open source or otherwise. RFC4015 has no such IPR. The community at the time felt strongly that the IPR meant RFC3522 shouldn't be standards track. Pasi developed F-RTO in part because of the IPR in RFC3522.

**Comment by huitema** [comment]
Note that as far as QUIC is concerned, detecting spurious RTO is completely straightforward. QUIC packets have unique numbers, any data that need to be repeated is carried in new packets, with numbers different from the initial transmission. The sender just has to record an ack of a packet that was considered lost, there is no requirement to implement F-RTO or other algorithms.

**Comment by larseggert** [comment]
Heads up: I plan to merge #111 soon, unless I hear objections. If you do object, please make a concrete rephrasing proposal (on the PR).

**Comment by markkukojo** [comment]
> > Do you think citing RFC 4015 for spurious RTOs is enough? This would mean we wouldn't provide any guidance for spurious detection of false fast retransmits.
> 
> I would suggest citing all the major mechanisms that major TCP stacks use in the real world to detect spurious retransmits, namely something like: "CUBIC implementations MAY use any of the following documented mechanisms for detecting spurious congestion events: Eifel [rfc3522] [rfc4015], F-RTO [rfc5682], DSACK [rfc3708]."
> 
> I believe their statuses are:
> 
> * Eifel rfc3522 (exp), rfc4015 (std)
> * F-RTO: rfc5682 (std)
> * DSACK: rfc3708 (exp)

Please note that RFC 4015 is a stds track *response* algorithm for spurious RTOs. The question was why the advise in RFC 4015 is wrong and why this draft would need to modify it as the current text does and thereby maybe requires to update RFC 4015, instead of pointing to RFC 4015 as the response algo for spurious RTOs and just giving the extra state that needs to be stored/restored?

**Comment by larseggert** [comment]
> Text for Slow Start section:

...

@markkukojo, so that entire paragraph should look like the following?
```
CUBIC MUST employ a slow-start algorithm, when *cwnd* is no more than
*ssthresh* {{!RFC5681}}. If CUBIC uses HyStart++
{{!I-D.ietf-tcpm-hystartplusplus}} for early exit from the initial
slow start, it may exit the slow start without incurring any packet
loss. Experimental alternatives to HyStart++ include hybrid slow
start {{HR08}}, a predecessor to HyStart++ that some CUBIC
implementations have used as the default for the last decade, and
limited slow start {{?RFC3742}}.
```

>     * I asked but did not get answer why CUBIC would be dependent on HyStart++ (use  SHOULD), i.e., what are the full implications that must be understood and carefully weighted before choosing not to implement HyStart++ with CUBIC? I don't see any such reasons, maybe I am missing something?

IIRC one pragmatic reason is that there is no other standards-track mechanism.

> I cannot provide text as I have expressed my strong opinion: -this draft must not provide an alternative stds track specification for response from spurious RTOs; we already have it in RFC 4015.

This sounds like you are arguing that text should be taken out of the document - which bits?

>     * A response to false fast rexmits should be a draft of its own because the solution is common to all congestion control algorithms, not in anyway specific to CUBIC, and the solution deserves much more attention and more careful design than what this draft currently provides.

Again, this sounds like you are arguing that text should be taken out of the document - which bits?

> I'd suggest to move the text in sec 4.9 to sec 5.3 (Difficult Environments): The first para extended with a brief intro for environments where and why spurious RTOs and false rexmits may occur. Pointers to results where the implications are shown. And pointing to detection algos and RFC 4015 for response to spurious RTOs + list what additional state in addition to ssthresh to store/restore.

First, I thought spurious RTOs and fals RTXs are not uncommon on general Internet paths. So I'm not sure that moving that text into a section called "difficult environments" is appropriate.

Second, it sounds like you don't just want to move the existing text, you are outlining an complete rewrite of what we have. It would be more actionable if you could propose a draft of the concrete text you think should be there (ideally, as a PR.)

**Comment by nealcardwell** [comment]
> a) blasts a full cwnd worth data at line speed,

That's a separate problem. Restart from idle does that as well. TCP implementations should use pacing to solve this problem of blasting full cwnds. Pacing is the only complete and general solution to blasting full cwnds, and it's orthogonal to CUBIC.

> b) does not consider adjusting dupack treshold to avoid further false rexmits,

That's an orthogonal loss recovery issue, and belongs in a separate document focused on loss recovery. For example, the RACK RFC covers this kind of issue. It doesn't make sense to put this in a CUBIC document, because the CUBIC document doesn't know whether 

> c) does not need to be prepared for a lying receiver that can easily make a fast rexmit to appear as a spurious rexmit, i.e., it is fine make it easy to fool the sender not to react to congestion at all?

Interactions between lying receivers and the detection of spurious loss recovery episodes are a loss recovery issue, not a congestion control issue, and belong in a loss recovery document.


> Sure, I am not objecting to specifying which variables need to be saved and
>  restored in addition to those that the existing RFCs already do and what
> is common to all CCs. But what I try to say is that this is the quite
> trivial part of the cc response to spurious rexmit.

Most of the issues you are bringing up are loss recovery issues, not congestion control issues, and should be in a separate document, IMHO. IMHO the 8312bis is a CC document, and should effectively say something like: "if your loss detection algorithm detects that a previous loss inference was spurious, then the CUBIC CC response is X".


> a) the TCP sender blasts a full restored cwnd worth of unnecessary rexmits
> at the bottleneck line speed,

If the loss recovery has been undone, the sender is not in a loss recovery state, and thus would not retransmit anything.

> b) does not adjust RTO and RTO calculation to avoid further spurious RTOs,

That's an orthogonal loss recovery issue about how to adjust the RTO calculation based on an undo even, and belongs in a separate document focused on loss recovery. 

> c) executes UNDO even if it received the acceptable Ack with ECE,

That's an orthogonal loss recovery issue about whether undo should happen, and belongs in a separate document focused on loss recovery.

**Comment by larseggert** [comment]
If we do this (1) would this address @markkukojo's issues and (2) would we need to cite RFC4015 and say something about it?

**Comment by goelvidhi** [comment]
No, I don't think we need to cite anything. Folks who would want to implement Eifel response would follow RFC 4015

**Comment by larseggert** [comment]
We added that section in response to #23, and there was quite a bit of discussion and support for it. It seems odd to go back on this now?

**Comment by markkukojo** [comment]
My sincere apologies again not being able to respond earlier (unfortunate force major for me).

About slow start start text in Sec 4.10:

I've already suggested text to replace the current one and explained the rationale. 
I try to repeat briefly. The current text is impossible; CUBIC or any other CC cannot replace RFC 5681 slow start by HyStart++ in general because HyStart++ is applicable to initial slow start only. It cannot be used during the slow start in RTO recovery! 

Please see draft-ietf-tcpm-hystartplusplus, Sec 4.3:
   An implementation SHOULD use HyStart++ only for the initial slow
   start (when ssthresh is at its initial value of arbitrarily high per
   [RFC5681]) and fall back to using traditional slow start for the
   remainder of the connection lifetime.

The problem with overshoot and the MD factor to use is present in the RTO recovery as well because a pkt loss may be detected before cwnd reaches ssthresh (even though HyStart++ suggest differently).
 
Please see shortly my reply to Bob in #85 where I clarify this further.

**Comment by markkukojo** [comment]
About spurious retransmissions and response to such events (Sec 4.9):

I agree with Vidhi that the issues and the response to spurious rexmits is not that trivial. I also agree with Neal that the issues related to response I listed belong to another document.

However, Neal suggests for some reason that those issues are loss recovery issues which sounds very weird to me. We are discussing spurious rexmit detection and response when such event is detected; when a spurious rexmit is detected there is no loss recovery by definition because detecting a spurious rexmit avoids/terminates loss recovery. 

Instead, this topic was extensively studied and considered at the time  of writing RFC 3522, RFC 5682, RFC 3708, and RFC 4015. The IETF decided to handle spurious rexmits by separating the detection and response in separate documents and it is the current IETF consensus as documented in these RFCs. The decision was to not include any response actions in the detection documents but to put all actions related to response in the response documents. These actions include congestion control actions and any other actions that were considered important for a sender to take.  This of course is not the only possibly way to organize RFCs for handling the issues, but it is what IETF  currently has agreed on. We may change this but it requires someone to volunteer rewriting all these documents.

The response actions are currently documented in RFC 4015 for detected spurious RTOs. AFAIK IETF has never accepted a standards track RFC that modifies existing congestion control and/or related important actions without appropriate experimental data to support such a change. AFAIK nobody has presented any data for discussion in TCPM wg that would support ignoring/modifying RFC 4015 response. Maybe chairs can clarify whether this is the case? 

IMO the only thing that this draft is subject to present w.r.t. response to a spurious RTO is the additional congestion control variables that needs to be stored/restored for CUBIC, that is,  W_max, K, epoch_start, and W_est.

I am happy to suggest text for this, if we can agree on this.

For spurious fast rexmits we do not have a response RFC for various reasons that I have tried to explain and do try to clarify here further.

First of all, again, it is not that trivial as this draft suggests and Neal seems to think. When a false rexmit is detected there are various different scenarios that require different handling and the response also depends on the detection method used. I believe that Neal mainly considers using DSACK-based detection? If DSACK is used, the fast recovery is over and all related ACKs have been received when the decision is made, i.e., there should not be any data segments in flight that were injected during the fast recovery. In this case the response is much more trivial and more closer to what this draft suggests. However,  using RFC 3708 to detect false rexmits and applying an UNDO CC response to *is not* ready for standard track for security reasons! There is no working solution for RFC 3708 algorithm against lying receivers, meaning that it is very easy to device a TCP receiver that avoids cwnd reductions in case of pkt loss! TCPM chairs would need to address this in their writeup and I would be suprised if such draft would pass SECDIR review or actually I would expect TCPM not to decide forwarding such proposal for publication.

If Eifel (RFC 3522) is used to detect false rexmits, the response must be more complex, because there might be a half a window worth spuriously rexmitted data segments in flight when Eifel detects the spurious false rexmit. That is, a trivial case where the reordering extent (distance) is small (just enough to trigger a false fast rexmit) and maybe one or a few segments only got reordered (delayed). In such a case, a new ACK arrives only almost immediately after the false fast  rexmit and maybe it is the only segment unnecessarily rexmitted. However, if segments are delayed such that they arrive at the receiver as the last segments in the current window and there is a relatively large chunk of segments reordered (maybe half of the window) and SACK is in use, the sender unnecessarily rexmits half a window of segments before Eifel detects that the fast rexmit was spurious. And, when Eifel detects it, it does not know what has 
happened to the rest of the window, i.e., whether it was unnecessarily rxmitted or if actually those segments were lost (or some of them). Therefore, the response algorithm must take this into account and it must not blindly restore the old value of cwnd. Unfortunately, the original Eifel paper did not properly cover different possible reordering cases (nor did we when Eifel RFCs were prepared if I recall it correctly). So, we do not have a proper response   algorithm for Eifel-detected false fast rexmits nor has anyone presented any data on how UNDO works with Eifel detection in various different scenarios.
Moreover, AFAIK there is no data ever gathered and presented on how the safe version of Eifel performs. Therefore, IMHO a response for spurious false rexmits is definitely not ready for standards track.

I think we all agree that the problem of spurious rexmits is important to solve and it should be properly. Therefore, IMO we should handle these in separate documents common to all CCs as we already did for spurious RTOs (which actually needs an update), not in this draft without any evidence presented to the wg for making the decision.

**Comment by goelvidhi** [comment]
> IMO the only thing that this draft is subject to present w.r.t. response to a spurious RTO is the additional congestion control variables that needs to be stored/restored for CUBIC, that is, W_max, K, epoch_start, and W_est.

@markkukojo I presented your objections on this topic at the TCPM IETF meeting today (still Nov 11th in PST) and I see two problems that are a bit tricky to solve.

1. The response described in RFC 4015 for setting cwnd is very conservative. It sets cwnd to flight size + min(IW, bytes_acked), which could be very low if the spurious detection happens at the tail of the window.
2. Spurious RTOs are less common than spurious fast retransmits in TCP. For QUIC, there are no spurious RTOs as the only way to detect losses is after receiving an ACK, which is equivalent to fast retransmits. So, QUIC implementations wouldn't be able to take advantage of spurious detection.

> AFAIK IETF has never accepted a standards track RFC that modifies existing congestion control and/or related important actions without appropriate experimental data to support such a change. 
I have not attended all IETF meetings in the past, so I can't say what data may or may not have been presented. But I can say for sure that there are many QUIC implementations (including ours) that are doing UNDO on detection of spurious retransmits and the improvement in throughput was significant for us to keep using it. If you'd like I can get an exact number of how many implementations that are widely deployed are doing this.

So, I am starting to lean towards this approach. How about we document how CUBIC should respond to spurious events due to, a. RTOs and
b. fast retransmits.

We can use RFC 4015's response for  a. and we can do the UNDO method for b. (backed by multiple QUIC deployments)

**Comment by nealcardwell** [comment]
@goelvidhi 
> How about we document how CUBIC should respond to spurious events due to,
> a. RTOs and
> b. fast retransmits.

This sounds good to me. Thanks, Vidhi!

@markkukojo said:
>  I believe that Neal mainly considers using DSACK-based detection?

As I noted,I'm referring to several different detection mechanisms:
"(FRTO, DSACKs, Eifel TS-based undo)". :-)

> There is no working solution for RFC 3708 algorithm against lying receivers

If we're going to stop work based on lying receivers, this is going to hold
up all TCP congestion control and loss recovery work unreasonably. As
has been clear since at least 1999 ("TCP Congestion Control with a
Misbehaving Receiver"), TCP has no crypto/nonce mechanism for
receivers to prove what they have received, so receivers can always
lie and claim they have received all packets, and thus cause the
sender to send as fast as the sender hardware and application will
allow, defeating just about any existing TCP congestion control or loss
recovery mechanism.

> So, we do not have a proper response algorithm for Eifel-detected false fast rexmits

I believe we do have such an algorithm, even if it's not documented
yet in an IETF document: Linux TCP has a proper response algorithm for
Eifel-detected false fast rexmits, which properly handles the case you
mention since 2013 (patch was by @yuchungcheng):

https://git.kernel.org/pub/scm/linux/kernel/git/netdev/net-next.git/commit/?id=7026b912f97d912476dff5465ed9a127be094208

> I think we all agree that the problem of spurious rexmits is important
> to solve and it should be properly. Therefore, IMO we should handle
> these in separate documents common to all CCs as we already did for
> spurious RTOs (which actually needs an update), not in this draft
> without any evidence presented to the wg for making the decision.

I think this passage is missing a key distinction. I think it is
useful to make a distinction between:

(a) How the TCP connection as a whole, and loss detection, and
detection of spurious retransmits should decide when to undo a
spurious loss detection event.

(b) What the congestion control module should do when the logic in (a)
decides that a loss detection event was spurious.

I continue to argue that it is useful and best for 8312bis to document
the (b) aspect of CUBIC. Partly because integrating the (b) aspect of
CUBIC into some future draft about (a) would be something of a
modularity violation, in that it would needlessly drag the details of
CUBIC state variables into what ideally should be a largely
congestion-control-agnostic document about (a).

**Comment by larseggert** [comment]
Could we drive this issue to closure, please?

**Comment by goelvidhi** [comment]
I will send out a PR in a day or two.
